airbyte-integrations/connectors/source-microsoft-onedrive/source_microsoft_onedrive/stream_reader.py

```diff
@@ -0,0 +1,178 @@
+        return ConfidentialClientApplication(
+            self.config.credentials.client_id,
+            authority=f"https://login.microsoftonline.com/{self.config.credentials.tenant_id}",
+            client_credential=self.config.credentials.client_secret,
+        )
+
+    @property
+    def client(self):
+        """Initializes and returns a GraphClient instance."""
+        if not self.config:
+            raise ValueError("Configuration is missing; cannot create the Office365 graph client.")
+        if not self._client:
+            self._client = GraphClient(self._get_access_token)
+        return self._client
+
+    def _get_access_token(self):
+        """Retrieves an access token for OneDrive access."""
+        scope = ["https://graph.microsoft.com/.default"]
+        refresh_token = self.config.credentials.refresh_token if hasattr(self.config.credentials, "refresh_token") else None
+
+        if refresh_token:
+            result = self.msal_app.acquire_token_by_refresh_token(refresh_token, scopes=scope)
+        else:
+            result = self.msal_app.acquire_token_for_client(scopes=scope)
+
+        if "access_token" not in result:
+            error_description = result.get("error_description", "No error description provided.")
+            raise MsalServiceError(error=result.get("error"), error_description=error_description)
+
+        return result
+
+
+class SourceMicrosoftOneDriveStreamReader(AbstractFileBasedStreamReader):
+    """
+    A stream reader for Microsoft OneDrive. Handles file enumeration and reading from OneDrive.
+    """
+
+    def __init__(self):
+        super().__init__()
+
+    @property
+    def config(self) -> SourceMicrosoftOneDriveSpec:
+        return self._config
+
+    @property
+    def one_drive_client(self) -> SourceMicrosoftOneDriveSpec:
+        return SourceMicrosoftOneDriveClient(self._config).client
+
+    @config.setter
+    def config(self, value: SourceMicrosoftOneDriveSpec):
+        """
+        The FileBasedSource reads and parses configuration from a file, then sets this configuration in its StreamReader. While it only
+        uses keys from its abstract configuration, concrete StreamReader implementations may need additional keys for third-party
+        authentication. Therefore, subclasses of AbstractFileBasedStreamReader should verify that the value in their config setter
+        matches the expected config type for their StreamReader.
+        """
+        assert isinstance(value, SourceMicrosoftOneDriveSpec)
+        self._config = value
+
+    def list_directories_and_files(self, root_folder):
+        """Enumerates folders and files starting from a root folder."""
+        drive_items = root_folder.children.get().execute_query()
+        found_items = []
+        for item in drive_items:
+            if item.is_file:
+                found_items.append(item)
+            else:
+                found_items.extend(self.list_directories_and_files(item))
```

Comment:
This logic is not building out paths for nested folders correctly - let's say I have this structure:
* root_folder
  * folder1
    * myfile.csv
  * folder2
    * myfile.csv

With the current schemes, both of these will have the name `myfile.csv`, so I can't use the glob to differentiate. The names should be `folder1/myfile.csv` and `folder2/myfile.csv` . This is how it works for Google Drive: https://github.com/airbytehq/airbyte/blob/136535d546ded551035b5251b7d3169fb21bb127/airbyte-integrations/connectors/source-google-drive/source_google_drive/stream_reader.py#L115

Reply:
It's working correctly. This part of code is not building path. It just collect files objects.

Reply:
fixed